### PR TITLE
Kh04

### PR DESCRIPTION
**Pull request by KhalidTheeb**
Base
 Test that </i> and other close tags work

Comments
 Test for --> outside a comment (it should just be normal text)
 Test combinations like  or other "nested" cases (should be one comment)
 Test  (should be one comment)

Paragraphs
 Test when the second <p> tag has different attributes from the first
 Test that parent pointers are set correctly

Script tags
 Test <script> tags with attributes (like <script defer>)

### CODE REVIEW

File: ch04/base.md [hunk 1]
```diff
@@ -37,75 +37,85 @@ resulting tree.
...
-	   <body>
-	     <div name1="value1" name2="value2">
-	       'text'
+        >>> test_parse("<div name1=value1 name2=value2>text</div")
+         <html>
+           <body>
+             <div name1="value1" name2="value2">
+               'text'
+
+Test that the italic `</i>` tag behaves correctly:
```
Comment (pavpanchekha) [inline ch04/base.md:93]: This isn't what we should test for: we should be testing that `</i>` in fact makes text stop being italic. So instead of `test_parse` we should parse it, make a `Layout` out of it, and print the `display_list`, which should show two words with the first one being italic but the second one not being italic. You can put it in the "Testing Layout" sections and model it off the first test in that section.

File: ch04/ex-comments.md [hunk 2]
```diff
@@ -94,3 +94,32 @@ There are some edge cases to take care of.
...
+
+
+Test case for handling text outside of comments
+
+    >>> test_parse("This is normal text --> following text should also be normal")
+     <html>
+       <body>
+         <this is="" normal="" text="" --="">
+           ' following text should also be normal'
+
```
Comment (pavpanchekha) [inline ch04/ex-comments.md:107]: Let's just delete this test. It probably *is* how the browser behaves right now, but this is pretty weird behavior so maybe it's better not to test for it.

File: ch04/ex-paragraphs.md [hunk 3]
```diff
@@ -54,3 +54,29 @@ Any tags that are open when encountering the second paragraph should be closed
          <p>
            <b>
              'world'
+
+
+Test when `<p>` tags have different attributes from each other
+
+    >>> test_parse("<p>This is a paragraph with <p>another nested paragraph inside.</p></p>")
```
Comment (pavpanchekha) [inline ch04/ex-paragraphs.md:61]: Don't put the closing `</p></p>` tags

Comment (pavpanchekha) [reply ch04/ex-paragraphs.md:61]: Also these `<p>` tags do not have different attributes from one another. They don't have any attributes at all!

File: ch04/ex-paragraphs.md [hunk 4]
```diff
@@ -54,3 +54,29 @@ Any tags that are open when encountering the second paragraph should be closed
...
+    >>> test_parse("<p>This is <b>paragraph one</b> with <p>nested paragraph two</p></p>")
+     <html>
+       <body>
+         <p>
+           'This is '
+           <b>
+             'paragraph one'
+           ' with '
+         <p>
+           'nested paragraph two'
```
Comment (pavpanchekha) [inline ch04/ex-paragraphs.md:81]: How is this testing for parent pointers?

File: ch04/ex-scripts.md [hunk 5]
```diff
@@ -65,3 +65,13 @@ The script should end only with a complete end script tag.
        <head>
          <script>
            ' "</scri" "pt>" '
+
+
+ Test `<script>` tags with attributes
+
+    >>> test_parse("<script defer src='my-script.js'>Should be treated as text</script>")
```
Comment (pavpanchekha) [inline ch04/ex-scripts.md:72]: Change the contents of the `<script>` tag to be `a<b`. This will test that the browser actually enters "script tag" mode where `<` and `>` are handled specially.

**Review by pavpanchekha — CHANGES_REQUESTED** [review]

File: ch04/base.md [hunk 6]
```diff
@@ -37,75 +37,87 @@ resulting tree.
...
+Parse HTML with italic text
+
+    >>> parser = browser.HTMLParser("<i>Italic</i> Normal.")
+    >>> tree = parser.parse()
+
+Create a Layout from the parsed tree
+
+    >>> lo = browser.Layout(tree)
+    >>> lo.display_list
+    [(13, 21.0, 'Italic', Font size=16 weight=normal slant=italic style=None), (125, 21.0, 'Normal.', Font size=16 weight=normal slant=roman style=None)]
```
Comment (pavpanchekha) [inline ch04/base.md:123]:
```suggestion
    >>> wbemocks.print_list(lo.display_list)
    (13, 21.0, 'Italic', Font size=16 weight=normal slant=italic style=None)
    (125, 21.0, 'Normal.', Font size=16 weight=normal slant=roman style=None)
```

File: ch04/ex-paragraphs.md [hunk 7]
```diff
@@ -54,3 +54,37 @@ Any tags that are open when encountering the second paragraph should be closed
...
+       <body>
+         <p style="font-family:foo">
+           '\nA\n'
+         <p class="diff">
+           '\nB\n'
+
+
+Test that parent pointers are set correctly
+
+    >>> test_parse("<p>This is <b>paragraph one</b> with <p>nested paragraph two</p></p>")
```
Comment (pavpanchekha) [inline ch04/ex-paragraphs.md:81]:
```suggestion
    >>> tree = browser.HTMLParser("<p id=one>This is <b>paragraph one</b> with <p id=two>nested paragraph two</p></p>").parse()
    >>> b_elt = tree.children[0].children[0].children[1]
    >>> b_elt.parent
    <p>
```
